cmd/argo/commands/cp.go

```diff
@@ -0,0 +1,145 @@
+			}
+
+			ctx, apiClient := client.NewAPIClient(cmd.Context())
+			serviceClient := apiClient.NewWorkflowServiceClient()
+			namespace := client.Namespace()
+			if len(copyArtifactArgs.namespace) > 0 {
+				namespace = copyArtifactArgs.namespace
+			}
+			workflow, err := serviceClient.GetWorkflow(ctx, &workflowpkg.WorkflowGetRequest{
+				Name:      workflowName,
+				Namespace: namespace,
+			})
+			if err != nil {
+				return fmt.Errorf("failed to get workflow: %w", err)
+			}
+
+			workflowName = workflow.Name
+			artifactSearchQuery := v1alpha1.ArtifactSearchQuery{
+				ArtifactName: copyArtifactArgs.artifactName,
+				TemplateName: copyArtifactArgs.templateName,
+				NodeId:       copyArtifactArgs.nodeId,
+			}
+			artifactSearchResults := workflow.SearchArtifacts(&artifactSearchQuery)
+
+			c := &http.Client{
+				Transport: &http.Transport{
+					TLSClientConfig: &tls.Config{
+						InsecureSkipVerify: argoServerOpts.InsecureSkipVerify,
+					},
+				},
+			}
+
+			basePath := filepath.Join(outputDir, namespace, workflowName)
+			for _, artifact := range artifactSearchResults {
+				nodeIdFolder := filepath.Join(basePath, artifact.NodeID, "outputs")
+				err = os.MkdirAll(nodeIdFolder, os.ModePerm)
+				if err != nil {
+					return fmt.Errorf("failed to create nodeId folder: %w", err)
+				}
+				key, err := artifact.GetKey()
+				if err != nil {
+					return fmt.Errorf("error getting key for artifact: %w", err)
+				}
+				err = getAndStoreArtifactData(namespace, workflowName, artifact.NodeID, artifact.Name, path.Base(key), nodeIdFolder, c, argoServerOpts)
+				if err != nil {
+					return fmt.Errorf("failed to get and store artifact data: %w", err)
+				}
+			}
+			return nil
+		},
+	}
+	command.Flags().StringVarP(&copyArtifactArgs.namespace, "namespace", "n", "", "namespace of workflow")
+	command.Flags().StringVar(&copyArtifactArgs.nodeId, "node-id", "", "id of node in workflow")
+	command.Flags().StringVar(&copyArtifactArgs.templateName, "template-name", "", "name of template in workflow")
+	command.Flags().StringVar(&copyArtifactArgs.artifactName, "artifact-name", "", "name of output artifact in workflow")
+	return command
+}
+
+func getAndStoreArtifactData(namespace string, workflowName string, nodeId string, artifactName string, fileName string, nodeIdFolder string, c *http.Client, argoServerArgs apiclient.ArgoServerOpts) error {
+	request, err := http.NewRequest("GET", fmt.Sprintf("%s/artifacts/%s/%s/%s/%s", argoServerArgs.GetURL(), namespace, workflowName, nodeId, artifactName), nil)
+	if err != nil {
+		return fmt.Errorf("failed to create request: %w", err)
+	}
+	request.Header.Set("Authorization", client.GetAuthString())
+	resp, err := c.Do(request)
+	if err != nil {
+		return fmt.Errorf("request failed with: %w", err)
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != 200 {
+		return fmt.Errorf("request failed %s", resp.Status)
+	}
+	artifactFilePath := filepath.Join(nodeIdFolder, fileName)
```

Comment:
The user will want control over the path. What did you think of my idea of a temptable path?

Reply:
By temptable path do you mean allowing a user to flatten the directory?

Reply:
I'm not sure. I think as a user, I've a few ways I want to use this.

Get all artifacts from several massive workflow. I'll invoke the CLI once per workflow. 

```bash
argo cp wf-1 ./out-1
argo cp wf-2 ./out-2
```

In that case, all the files should be put into the directory in such a way they are all uniquely named. E.g. `{nodeId}/outputs/{artifactName}`

Get a single artifacts from a workflow. 

```bash
argo cp wf-1 . -1 --template=output-tmpl --filename=output.json
```

Get a few artifacts from a workflow, where each template is only used once:

```bash
argo cp wf-1 . -1 --template=output-tmpl --filename='{templateName}/{artifactName}'
```

Get a artifacts from a workflow, but each has different key (filename=basename(key):

```bash
argo cp wf-1 . -1 --template=output-tmpl --filename='{filename}'
```

Any I missed?

Reply:
I see what you're saying. I think given the direction you're talking about, it might make sense to fully commit to exhaustive customizability for path construction to keep things as simple as possible for users (would probably be easier for a user to understand than case-by-case).

We could have 4 main variables - `workflowName`, `nodeId`, `templateName`, `artifactName`. A user could use the --filename (although if we're using it as a path, --path would make more sense) to define a path structure. Eg:

```
--path=`{templateName}/{artifactName}`
--path=`{workflowName}/{nodeId}/{artifactName}`
--path=`{workflowName}`
etc...
```
This would allow a user to create a fully customized path based on whatever their requirements are. The default would be `{workflowName}/{nodeId}/outputs/{artifactName}`.

All the other filter flags would still be honored as the path would only be made for artifacts that are being copied. The artifact file would be stored in whatever the constructed folder is.

Do you think this might be a good approach for customizable paths?

Reply:
You’re missing one variable `filename`, which is `basename(key)`. But yes, and `—path` does seem more correct.